python/ql/src/semmle/python/security/strings/External.qll

```diff
@@ -181,6 +185,58 @@ private predicate urlparse(ControlFlowNode fromnode, CallNode tonode) {
     )
 }
 
+private predicate parse_qs(ControlFlowNode fromnode, CallNode tonode) {
+    // This could be implemented as `exists(FunctionValue` without the explicit six part,
+    // but then our tests will need to import +100 modules, so for now this slightly
+    // altered version gets to live on.
+    exists(Value parse_qs |
+        (
+            parse_qs = Value::named("six.moves.urllib.parse.parse_qs")
+            or
+            // Python 2
+            parse_qs = Value::named("urlparse.parse_qs")
+            or
+            // Python 2 deprecated version of `urlparse.parse_qs`
+            parse_qs = Value::named("cgi.parse_qs")
+            or
+            // Python 3
+            parse_qs = Value::named("urllib.parse.parse_qs")
+        ) and
+        tonode = parse_qs.getACall() and
+        (
+            tonode.getArg(0) = fromnode
+            or
+            tonode.getArgByName("qs") = fromnode
+        )
+    )
+}
+
+private predicate parse_qsl(ControlFlowNode fromnode, CallNode tonode) {
+    // This could be implemented as `exists(FunctionValue` without the explicit six part,
+    // but then our tests will need to import +100 modules, so for now this slightly
+    // altered version gets to live on.
+    exists(Value parse_qsl |
+        (
+            parse_qsl = Value::named("six.moves.urllib.parse.parse_qsl")
+            or
+            // Python 2
+            parse_qsl = Value::named("urlparse.parse_qsl")
+            or
+            // Python 2 deprecated version of `urlparse.parse_qsl`
+            parse_qsl = Value::named("cgi.parse_qsl")
+            or
+            // Python 3
+            parse_qsl = Value::named("urllib.parse.parse_qsl")
```

Comment:
We should maybe factor out all of this "Python 2, Python 2 (deprecated), Python 3" logic in a shared predicate `urllib_method(string name)` to avoid copy-pasting it everywhere?

Reply:
I see your point and I sort of agree. However, notice that the _Python 2 (deprecated)_ is only for `parse_qsl` and `parse_qs`.

We could have a `urllib_parse_method` helper predicate to encapsulate six/py2/py3. If we had 10 special cases I would probably be totally on board with that idea, but personally I don't feel a strong urge to do anything about it.

If you feel strongly about it, I'd be happy to do it. I'm probably a 2 out of 10 on [The Sliding Scale of Giving a Fuck](https://capwatkins.com/blog/the-sliding-scale-of-giving-a-fuck).

Reply:
Nah, that's fine. I don't feel strongly about it.